filesmanager/filesmanager.py

```diff
@@ -570,15 +593,47 @@ def upload_file_to_directory(self, file, target_directory):
             from cms.djangoapps.contentstore.asset_storage_handler import \
                 update_course_run_asset  # pylint: disable=import-outside-toplevel
 
-        file_object = self.temporary_uploaded_files.pop(file.get("name"), None)
+        file_object = self.find_temporary_file(file.get("name"))
         file_path, name = file.get("path"), file.get("name")
         metadata = file.get("metadata", {})
 
-        if file_object:
+        if file_object and not metadata:
+            # New file uploaded from the Files Manager
             file_path, name = self.generate_content_path(file_path, file_object.filename)
-            file_object.file._set_name(name) # pylint: disable=protected-access
+            file_object.file._set_name(self.generate_asset_name(file_path)) # pylint: disable=protected-access
             content = update_course_run_asset(self.course_id, file_object.file)
             metadata = self.get_asset_json_from_content(content)
+        elif not metadata.get("uploaded_at"):
+            # File was uploaded from the course assets. Just create a link
+            if "files-" not in metadata.get("asset_key"):
+                file_path, name = self.generate_content_path(file_path, metadata.get("display_name"))
+                metadata["from"] = metadata.get("asset_key")
+                self.source_keys[metadata.get("asset_key")] = metadata.get("asset_key")
+            else:
+                # If the file is bring back to the Unpublished folder and was not uploaded from the course assets
+                # mark the file as Unpublished again and do not upload it to the course assets
+                if "Unpublished" in file_path and not metadata.get("uploaded_at"):
+                    if metadata.get("from"):
+                        del self.source_keys[metadata.get("from")]
+                    return
+        else:
+            # File uploaded to files-manager and then moved to a different directory
+            file_path, name = self.generate_content_path(file_path, name)
+            internal_name = self.generate_asset_name(file_path)
+            if "Unpublished" in file_path and not metadata.get("uploaded_at"):
+                if metadata.get("from"):
+                    del self.source_keys[metadata.get("from")]
+                return
+            if not internal_name in metadata.get("asset_key"):
+                # File was moved from a different directory
+                source_asset_key = metadata.get("asset_key")
+                memory_file = self.generate_memory_file_for_asset(metadata)
+                memory_file._set_name(internal_name) # pylint: disable=protected-access
+                content = update_course_run_asset(self.course_id, memory_file)
+                metadata = self.get_asset_json_from_content(content)
+                self.delete_asset(source_asset_key)
+        if not metadata:
+            raise Exception("Metadata not found")
```

Comment:
Can this have fewer nested-ifs? Can we try refactoring? It's a bit confusing even with the in-line comments

Reply:
Other attempts to refactor this code have led to really weird bugs.